contracts/utils/StaticDelegateCall.sol

```diff
@@ -0,0 +1,38 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+
+/**
+ * @dev Provides emulation of static delegatecalls for implementing view-only
+ * functions that execute another contract's code with the calling contract's
+ * context.
+ */
+abstract contract StaticDelegateCall {
+    /**
+     * @dev Perform a static delegate call to the specified `target` with
+     * calldata `data`. Returns success boolean and return data.
+     */
+    function _staticDelegateCall(address target, bytes memory data) internal view returns (bool, bytes memory) {
+        /* This staticcall always reverts, so we can ignore the success boolean */
+        (, bytes memory returndata) = address(this).staticcall(
+            abi.encodeCall(this.delegateCallAndRevert, (target, data))
+        );
+        return abi.decode(returndata, (bool, bytes));
+    }
+
+    /**
+     * @dev Simulate a delegatecall to the specified `target` with calldata
+     * `data` by reverting with the results. The resulting success boolean and
+     * return data are ABI-encoded in the revert data.
+     *
+     * Inspired by simulateAndRevert() from Gnosis' StorageSimulation contract.
+     */
+    function delegateCallAndRevert(address target, bytes memory data) external {
```

Comment:
Am I understanding correctly that this file is an `abstract contract` because this function needs to be external?

If that's the case I'd make this call internally and let the users expose it, so we can make this a library.
Note that we generally avoid libraries with external calls, so [library linking](https://docs.soliditylang.org/en/latest/using-the-compiler.html#library-linking) is not needed.

I understand making it `internal` conflicts with the need for `_staticDelegateCall` to call `delegateCallAndRevert`, but still makes sense to me that an implementer would be required to expose it.

Reply:
I'm a little unsure of how this would work as a library -- which target would `_staticDelegateCall()` encode? Would the callsite get passed in somehow?

Reply:
What I see is that these two functions are strictly tied together while they belong to the same action, any reason why they should be decoupled?

What I'm saying is that, if we can couple them together into a single function we might be able to either publish it as an `internal` function in a library or as an `external` one in an `abstract contract`, similar to what we do with Multicall:

```solidity
    function staticDelegateCall(
        address target,
        bytes memory data,
        bool delegated
    ) external returns (bool, bytes memory) {
        return
            delegated
                ? _delegateCallAndRevert(target, data)
                : _staticDelegateCall(target, data);
    }

    function _staticDelegateCall(
        address target,
        bytes memory data
    ) private view returns (bool, bytes memory) {
        /* This staticcall always reverts, so we can ignore the success boolean */
        (, bytes memory returndata) = address(this).staticcall(
            abi.encodeCall(this.staticDelegateCall, (target, data, true))
        );
        return abi.decode(returndata, (bool, bytes));
    }

    function _delegateCallAndRevert(
        address target,
        bytes memory data
    ) private returns (bool, bytes memory) {
        require(
            msg.sender == address(this),
            "StaticDelegateCall: caller must be self"
        );
        (bool success, bytes memory returndata) = target.delegatecall(data);
        bytes memory result = abi.encode(success, returndata);
        assembly {
            revert(add(result, 0x20), mload(result))
        }
    }
```

Although I'm not a big fan of the pattern, it makes clear what to use if we export this as an abstract contract.
Does something like this make sense to you?

Reply:
As a clarification regarding coupling -- the original coupling is necessary to have a view entry point (using `staticcall`) and a non-view executor (using `delegatecall`).

I'm trying to understand this approach -- are you thinking that the `external staticDelegateCall()` would go into `StaticDelegateCall` abstract contract, while `internal _staticDelegateCall()` and `internal _delegateCallAndRevert()` would go into the existing `Address` library, and the `StaticDelegateCall` abstract contract would then use the `Address` library as `Multicall` does?

I'm not sure this approach makes things cleaner overall, because the helpers `_staticDelegateCall()` and `_delegateCallAndRevert()` aren't useful on their own, and this arrangement imposes a few additional costs:
1) The footprint grows from two functions in a single abstract contract, to three functions spread across an abstract contract and a library, while still being coupled
2) There is an additional trampoline `_staticDelegateCall()` -> `staticDelegateCall()` -> `_delegateCallAndRevert()`
3) The `delegated` boolean parameter in `staticDelegateCall()` is not necessary and potentially misleading

There are also may be a few things missing with the posted implementation, i.e. `_staticDelegateCall()` and `_delegateCallAndRevert()` need to be at least internal for it to work, but maybe I'm not understanding where these functions are supposed to live.

Unless you think there may be another use case for `_delegateCallAndRevert()` in `Address`, I still think the original implementation is cleaner, easier to use, and easier to understand.

Reply:
I agree that the original implementation is cleaner. Even though this proposal is able to put all the logic inside a single external `staticDelegateCall` function, it does this at the expense of adding an extraneous boolean argument that is only used for internal logic. Additionally, the API is compromised because the user has to use `this.staticDelegateCall(..., false)`. If the end goal is to provide functionality to be used internally in a contract, an internal `_staticDelegateCall(...)` function is the best fit.